File: app/src/figures/compute-law-fit.ts
// Two fits of the compute-only scaling law on the same measured points, and
// how far their forecasts separate outside the measured range. The law is the
// field-map chapter's
//
//   L(C) = L∞ + A C^(−α),
//
// fitted twice: with L∞ = 0 (the pure power law of Kaplan et al. 2020) and
// with the floor L∞ fitted. The points are the compute frontier of the
// Chinchilla runs reconstructed by Besiroglu et al. (2024) from Figure 4 of
// Hoffmann et al. (2022): at each of nine IsoFLOP budgets, the lowest loss of a
// parabola in log N through the runs at that budget. The reader picks the
// largest budget in the fit (the larger ones are held out) and a forecast
// budget. Every fit and band is precomputed by
// tools/figure-data/compute-law-fit.py into data/compute-law-frontier.ts; the
// module evaluates the closed form and interpolates the band.

import { defineFigure, type Lang, type State } from "./types.ts";
import { svg, el, text, g, linePath } from "./lib/svg.ts";
import { C, TYPE } from "./lib/theme.ts";
import { log } from "./lib/scale.ts";
import { axis, axisHeight } from "./lib/axis.ts";
import { legend } from "./lib/legend.ts";
import { placeLabels, drawLabels, lineObstacles, textWidth, type Box, type LabelRequest } from "./lib/labels.ts";
import { fixed, sig, tpl } from "./lib/format.ts";
import { sci, pow10 } from "./lib/notation.ts";
import { wrapLines } from "./lib/wrap-lines.ts";
import { GOPHER_BUDGET } from "./data/scaling-laws.ts";
import { GRID, POINTS, FITS, type CutoffFit, type FormFit } from "./data/compute-law-frontier.ts";

const labels = {
  en: {
    title: "Two fits of the compute law, inside and outside the measured range",
    panel: "Lowest loss at each compute budget, and two fits of L(C)",
    xC: "training compute C (FLOP)",
    yL: "loss L (log scale)",
    fitted: "fitted budgets",
    heldOut: "held out",
    pure: "pure power law, L∞ = 0",
    floor: "fitted floor L∞",
    lgFit: "frontier point in the fit",
    lgHeld: "held-out point",
    lgBand: "80% bootstrap band",
    floorLine: "L∞ = {v}",
    summary: "Fit on the {k} budgets up to {c} FLOP; {h:larger budget/larger budgets} held out",
    summaryAll: "Fit on all {k} budgets up to {c} FLOP; nothing held out",
    coefPure: "A = {A}, α = {a}",
    coefFloor: "L∞ = {Li}, A = {A}, α = {a}",
    atPure: "at {c} FLOP: L = A C^−α = {v}",
    atFloor: "at {c} FLOP: L = L∞ + A C^−α = {Li} + {r} = {v}",
    held: "80% band {lo} to {hi}; median error on the held-out budgets {e}",
    heldNone: "80% band {lo} to {hi}; no held-out budgets to check",
    gap: "The two fits differ by at most {a} over the fitted budgets and by {b} at {c} FLOP.",
    source: "Points: the lowest loss of a parabola through the runs at each IsoFLOP budget, from Hoffmann et al. (2022) Figure 4 as reconstructed by Besiroglu et al. (2024), loss read to about 0.01. Bands: 10th to 90th percentile over 1,000 bootstrap resamples of the runs.",
    describe: "Fitted on the {k} budgets up to {c} FLOP, the pure power law forecasts L = {vp} at {t} FLOP and the fit with a floor L∞ = {li} forecasts {vf}, a gap of {b}; over the fitted budgets they differ by at most {a}. {held}",
    heldDesc: "On the {h:held-out budget/held-out budgets} the median error is {ep} for the pure law and {ef} with the floor.",
    heldDescNone: "No budgets are held out.",
  },
  zh: {
    title: "算力扩展律的两种拟合：测量范围之内与之外",
    panel: "各算力预算下的最低损失，以及 L(C) 的两种拟合",
    xC: "训练算力 C（FLOP）",
    yL: "损失 L（对数坐标）",
    fitted: "参与拟合的预算",
    heldOut: "留出",
    pure: "纯幂律，L∞ = 0",
    floor: "拟合下限 L∞",
    lgFit: "参与拟合的前沿点",
    lgHeld: "留出的前沿点",
    lgBand: "80% 自助法区间",
    floorLine: "L∞ = {v}",
    summary: "用不超过 {c} FLOP 的 {k} 个预算拟合；留出 {h} 个更大的预算",
    summaryAll: "用不超过 {c} FLOP 的全部 {k} 个预算拟合；没有留出预算",
    coefPure: "A = {A}，α = {a}",
    coefFloor: "L∞ = {Li}，A = {A}，α = {a}",
    atPure: "{c} FLOP 处：L = A C^−α = {v}",
    atFloor: "{c} FLOP 处：L = L∞ + A C^−α = {Li} + {r} = {v}",
    held: "80% 区间 {lo} 至 {hi}；留出预算上的中位误差 {e}",
    heldNone: "80% 区间 {lo} 至 {hi}；没有留出预算可供检验",
    gap: "在参与拟合的预算上，两种拟合最多相差 {a}；在 {c} FLOP 处相差 {b}。",
    source: "数据点：每个等算力预算下，用抛物线拟合该预算的各次训练，取其最低损失；训练数据来自 Hoffmann 等人（2022）图 4，由 Besiroglu 等人（2024）重建，损失精度约为 0.01。区间：对训练结果做 1,000 次自助法重采样，取第 10 至第 90 百分位。",
    describe: "用不超过 {c} FLOP 的 {k} 个预算拟合时，纯幂律在 {t} FLOP 处预测 L = {vp}，带下限 L∞ = {li} 的拟合预测 {vf}，两者相差 {b}；在参与拟合的预算上最多相差 {a}。{held}",
    heldDesc: "在 {h} 个留出预算上，纯幂律的中位误差为 {ep}，带下限的拟合为 {ef}。",
    heldDescNone: "没有留出预算。",
  },
};

const CUTS = ["1e20", "3e20", "6e20", "1e21", "3e21"] as const;
type Cut = (typeof CUTS)[number];
type P = { fitTo: Cut; target: number };
type Form = "pure" | "floor";

const C_DOMAIN: [number, number] = [1e18, 1e25];
const L_DOMAIN: [number, number] = [1.2, 3.4];
const L_TICKS = [1.2, 1.4, 1.6, 1.8, 2, 2.2, 2.4, 2.6, 2.8, 3, 3.2];
const COLOR: Record<Form, string> = { pure: C.c1, floor: C.c2 };

const fitOf = (cut: Cut): CutoffFit => FITS.find((f) => f.budget === Number(cut))!;

// The forecast budget; the slider's 23.76 is the Chinchilla and Gopher budget.
const targetOf = (t: number) => (Math.abs(t - Math.log10(GOPHER_BUDGET)) < 0.006 ? GOPHER_BUDGET : 10 ** t);

const lawAt = (f: FormFit, c: number) => f.Linf + f.A * c ** -f.alpha;

// Linear interpolation in log10 C of a band on GRID.
function band(arr: readonly number[], lc: number): number {
  const step = GRID[1] - GRID[0];
  const i = Math.max(0, Math.min(GRID.length - 2, Math.floor((lc - GRID[0]) / step)));
  const u = (lc - GRID[i]) / step;
  return arr[i] * (1 - u) + arr[i + 1] * u;
}

function forecast(f: FormFit, c: number) {
  const lc = Math.log10(c);
  return { v: lawAt(f, c), red: f.A * c ** -f.alpha, lo: band(f.lo, lc), hi: band(f.hi, lc) };
}

// Largest difference between the two fits over the fitted budgets.
function gapInside(fit: CutoffFit): number {
  let mx = 0;
  const l0 = Math.log10(POINTS[0].C), l1 = Math.log10(fit.budget);
  for (let lc = l0; lc <= l1 + 1e-9; lc += 0.02) mx = Math.max(mx, Math.abs(lawAt(fit.pure, 10 ** lc) - lawAt(fit.floor, 10 ** lc)));
  return mx;
}

// A coefficient as the readout writes it: 42, 1,460, 1.33 × 10⁴.
const coef = (v: number) => (v >= 1e4 ? sci(v, 3) : sig(v, 3));

function describe(st: State<P>, lang: Lang): string {
  const L = labels[lang];
  const fit = fitOf(st.p.fitTo);
  const c = targetOf(st.p.target);
  const held = POINTS.length - fit.points;
  return tpl(L.describe, {
    k: fit.points, c: sci(fit.budget), t: sci(c),
    vp: fixed(lawAt(fit.pure, c), 2), vf: fixed(lawAt(fit.floor, c), 2), li: fixed(fit.floor.Linf, 2),
    b: fixed(Math.abs(lawAt(fit.floor, c) - lawAt(fit.pure, c)), 2), a: fixed(gapInside(fit), 3),
    held: held
      ? tpl(L.heldDesc, { h: held, ep: fixed(fit.pure.errHeld ?? 0, 3), ef: fixed(fit.floor.errHeld ?? 0, 3) })
      : L.heldDescNone,
  });
}

function render(st: State<P>, lang: Lang): string {
  const L = labels[lang];
  const p = st.p;
  const w = st.w;
  const narrow = w < 480;
  const fs = narrow ? TYPE.body : TYPE.small;
  const fit = fitOf(p.fitTo);
  const k = fit.points - 1; // index of the largest fitted budget
  const c = targetOf(p.target);
  const clip = `${st.uid}-clip`;
  const parts: string[] = [];

  // ---- plot
  const x0 = narrow ? 38 : 42, x1 = w - 8;
  const head = wrapLines(L.panel, TYPE.label, w - 6);
  const top = 44 + (head.length - 1) * 18;
  const h = narrow ? 230 : 270, bot = top + h;
  const x = log(C_DOMAIN, [x0, x1]);
  const y = log(L_DOMAIN, [bot, top]);
  parts.push(el("defs", {}, el("clipPath", { id: clip }, el("rect", { x: x0, y: top, width: x1 - x0, height: h }))));
  head.forEach((ln, i) => parts.push(text(0, 14 + i * 18, ln, { "font-size": TYPE.label, class: "fig-t-strong" })));

  // The fitted range runs to halfway (in log C) to the first held-out budget.
  const edge = k < POINTS.length - 1 ? Math.sqrt(POINTS[k].C * POINTS[k + 1].C) : POINTS[k].C * 1.4;
  const ex = x(edge);
  parts.push(el("rect", { x: x0, y: top, width: ex - x0, height: h, fill: C.panel }));
  parts.push(axis({ scale: x, orient: "bottom", at: bot, grid: [top, bot], minor: true, title: L.xC, format: pow10, size: fs }));
  parts.push(axis({ scale: y, orient: "left", at: x0, grid: [x0, x1], ticks: narrow ? L_TICKS.filter((_, i) => i % 2 === 0) : L_TICKS, title: L.yL, format: (v) => fixed(v, 1), size: fs }));

  const obstacles: Box[] = [];
  const clipped: string[] = [];
  // The fitted floor, as a horizontal reference.
  const fy = y(fit.floor.Linf);
  clipped.push(el("line", { x1: x0, x2: x1, y1: fy, y2: fy, stroke: C.ink3, "stroke-width": 1, "stroke-dasharray": "5 4" }));
  obstacles.push(...lineObstacles([[x0, fy], [x1, fy]]));
  // Bands, then curves: solid over the fitted budgets, dashed beyond them.
  const forms: Form[] = ["pure", "floor"];
  for (const form of forms) {
    const f = fit[form];
    const lo = GRID.map((lc, i) => [x(10 ** lc), y(f.lo[i])] as [number, number]);
    const hi = GRID.map((lc, i) => [x(10 ** lc), y(f.hi[i])] as [number, number]);
    clipped.push(el("path", { d: linePath([...lo, ...hi.reverse()]) + "Z", fill: COLOR[form], "fill-opacity": 0.22 }));
  }
  const curve = (f: FormFit, c0: number, c1: number) => {
    const out: Array<[number, number]> = [];
    for (let lc = Math.log10(c0); lc <= Math.log10(c1) + 1e-9; lc += 0.05) out.push([x(10 ** lc), y(lawAt(f, 10 ** lc))]);
    return out;
  };
  for (const form of forms) {
    const inside = curve(fit[form], C_DOMAIN[0], edge);
    const beyond = curve(fit[form], edge, C_DOMAIN[1]);
    clipped.push(el("path", { d: linePath(inside), fill: "none", stroke: COLOR[form], "stroke-width": 2.2 }));
    clipped.push(el("path", { d: linePath(beyond), fill: "none", stroke: COLOR[form], "stroke-width": 2.2, "stroke-dasharray": "6 4" }));
    obstacles.push(...lineObstacles([...inside, ...beyond].filter(([, py]) => py >= top && py <= bot)));
  }
  parts.push(g({ "clip-path": `url(#${clip})` }, ...clipped));

  // Frontier points: filled in the fit, hollow when held out. A click on a
  // point makes it the largest fitted budget.
  POINTS.forEach((pt, i) => {
    const px = x(pt.C), py = y(pt.L);
    const held = i > k;
    parts.push(held
      ? el("circle", { cx: px, cy: py, r: 3.6, fill: C.paper, stroke: C.ink, "stroke-width": 1.5 })
      : el("circle", { cx: px, cy: py, r: 3.6, fill: C.ink }));
    const cut = CUTS.find((cc) => Number(cc) === pt.C);
    if (cut) parts.push(el("circle", { cx: px, cy: py, r: 10, fill: "transparent", "data-fig-set": `fitTo=${cut}`, class: "fig-hit" }));
    obstacles.push({ x0: px - 5, y0: py - 5, x1: px + 5, y1: py + 5 });
  });

  // Forecast: a vertical guide, then each fit's value and band at the target.
  const tx = x(c);
  parts.push(el("line", { x1: tx, x2: tx, y1: top, y2: bot, stroke: C.ink2, "stroke-width": 1, "stroke-dasharray": "2 3" }));
  obstacles.push(...lineObstacles([[tx, top], [tx, bot]]));
  const reqs: LabelRequest[] = [];
  const fc = { pure: forecast(fit.pure, c), floor: forecast(fit.floor, c) };
  for (const form of forms) {
    const v = fc[form];
    const ty = y(v.v), tlo = y(v.lo), thi = y(v.hi);
    parts.push(el("line", { x1: tx, x2: tx, y1: tlo, y2: thi, stroke: C.ink, "stroke-width": 2 }));
    for (const yy of [tlo, thi]) parts.push(el("line", { x1: tx - 4, x2: tx + 4, y1: yy, y2: yy, stroke: C.ink, "stroke-width": 2 }));
    parts.push(el("circle", { cx: tx, cy: ty, r: 5, fill: COLOR[form], stroke: C.paper, "stroke-width": 2 }));
    obstacles.push({ x0: tx - 7, y0: Math.min(thi, ty - 6), x1: tx + 7, y1: Math.max(tlo, ty + 6) });
    reqs.push({ x: tx, y: ty, text: fixed(v.v, 2), size: fs, sides: ["right", "left", "above-right", "below-right", "above-left", "below-left"], gap: 10, priority: 5, attrs: { class: "fig-t-halo fig-t-num" } });
  }
  // The floor is named at its left end, where the curves are still far above it.
  reqs.push({ x: x0 + 4, y: fy, text: tpl(L.floorLine, { v: fixed(fit.floor.Linf, 2) }), size: fs, sides: ["above-right", "below-right"], gap: 6, priority: 3, attrs: { class: "fig-t-halo fig-t-soft fig-t-num" } });
  // Region names along the bottom of the plot, below the curves.
  const regionY = bot - 8;
  if (ex - x0 > textWidth(L.fitted, fs) + 12) reqs.push({ x: x0 + 6, y: regionY - fs * 0.35, text: L.fitted, size: fs, sides: ["right"], gap: 0, priority: 4, attrs: { class: "fig-t-muted" } });
  if (k < POINTS.length - 1) reqs.push({ x: ex + 6, y: regionY - fs * 0.35, text: L.heldOut, size: fs, sides: ["right"], gap: 0, priority: 4, attrs: { class: "fig-t-muted" } });
  const placed = placeLabels(reqs, { x0: x0 + 2, y0: top + 2, x1: x1 - 2, y1: bot - 2 }, obstacles);
  parts.push(drawLabels(placed.placed));

  // Legend under the axis.
  const lg = legend([
    { label: L.pure, swatch: { kind: "line", stroke: COLOR.pure } },
    { label: L.floor, swatch: { kind: "line", stroke: COLOR.floor } },
    { label: L.lgFit, swatch: { kind: "dot", fill: C.ink } },
    { label: L.lgHeld, swatch: { kind: "rect", fill: C.paper, stroke: C.ink } },
    { label: L.lgBand, swatch: { kind: "rect", fill: C.ink3, opacity: 0.35 } },
  ], 0, bot + axisHeight(true, fs) + 6, w, fs);
  parts.push(lg.svg);

  // ---- readout: each fit's terms and its forecast, then the gap
  let yy = bot + axisHeight(true, fs) + 6 + lg.height + 22;
  const rp: string[] = [];
  const line = (s: string, cls: string, size: number, x = 0, after = 0) => {
    for (const part of wrapLines(s, size, w - x)) { rp.push(text(x, yy, part, { "font-size": size, class: cls || undefined })); yy += size + 5; }
    yy += after;
  };
  const held = POINTS.length - fit.points;
  line(held ? tpl(L.summary, { k: fit.points, c: sci(fit.budget), h: held }) : tpl(L.summaryAll, { k: fit.points, c: sci(fit.budget) }), "fig-t-strong", TYPE.body, 0, 8);
  const ind = 26;
  for (const form of forms) {
    const f = fit[form];
    const v = fc[form];
    rp.push(el("line", { x1: 0, x2: 18, y1: yy - 4, y2: yy - 4, stroke: COLOR[form], "stroke-width": 2.5, "stroke-linecap": "round" }));
    const law = form === "pure"
      ? tpl(L.coefPure, { A: coef(f.A), a: sig(f.alpha, 3) })
      : tpl(L.coefFloor, { Li: fixed(f.Linf, 3), A: coef(f.A), a: sig(f.alpha, 3) });
    if (narrow) {
      line(L[form], "fig-t-strong", TYPE.body, ind);
      line(law, "fig-t-num", TYPE.body, ind);
    } else {
      rp.push(text(ind, yy, L[form], { "font-size": TYPE.body, class: "fig-t-strong" }));
      rp.push(text(ind + 190, yy, law, { "font-size": TYPE.body, class: "fig-t-num" }));
      yy += TYPE.body + 5;
    }
    const at = form === "pure"
      ? tpl(L.atPure, { c: sci(c), v: fixed(v.v, 3) })
      : tpl(L.atFloor, { c: sci(c), Li: fixed(f.Linf, 3), r: fixed(v.red, 3), v: fixed(v.v, 3) });
    line(at, "fig-t-num", TYPE.body, ind);
    const bandVars = { lo: fixed(v.lo, 3), hi: fixed(v.hi, 3) };
    line(f.errHeld == null ? tpl(L.heldNone, bandVars) : tpl(L.held, { ...bandVars, e: fixed(f.errHeld, 3) }), "fig-t-muted fig-t-num", fs, ind, 10);
  }
  line(tpl(L.gap, { a: fixed(gapInside(fit), 3), b: fixed(Math.abs(fc.floor.v - fc.pure.v), 2), c: sci(c) }), "", TYPE.body, 0, 8);
  line(L.source, "fig-t-muted", fs, 0);
  parts.push(g({ class: "fig-readout" }, ...rp));
  return svg(w, yy + 2, describe(st, lang), ...parts);
}

export default defineFigure({
  name: "compute-law-fit",
  title: { en: labels.en.title, zh: labels.zh.title },
  labels,
  params: {
    fitTo: {
      kind: "choice", control: "buttons", label: { en: "Fit on budgets up to", zh: "拟合所用的最大预算" }, default: "1e20",
      options: CUTS.map((cut) => ({ value: cut, label: { en: `${sci(Number(cut), 1)} FLOP`, zh: `${sci(Number(cut), 1)} FLOP` } })),
    },
    target: {
      kind: "range", label: { en: "Forecast budget, log₁₀ FLOP", zh: "预测预算，log₁₀ FLOP" }, min: 19, max: 25, step: 0.01, default: 23.76,
      marks: [{ value: 23.76, label: { en: "Chinchilla 70B", zh: "Chinchilla 70B" } }],
    },
  },
  render,
  describe,
});

File: tools/figure-data/compute-law-fit.py
# Regenerate app/src/figures/data/compute-law-frontier.ts: the compute
# frontier of the Chinchilla training runs and two compute-only fits of it,
# for the compute-law-fit figure of the field-map chapter.
#
# The chapter writes the compute-only law with a constant floor,
#
#   L(C) = L_inf + A C^(-alpha),
#
# and notes that the compute law of Kaplan et al. (2020) is the pure power law,
# L_inf = 0. This script fits both forms to the same measured points and
# records how far their forecasts separate outside the measured range.
#
# Data: the training runs reconstructed by Besiroglu et al. (2024) from Figure 4
# of Hoffmann et al. (2022), epoch-research/analyzing-chinchilla,
# data/svg_extracted_data.csv at commit 92258837425e1b5f2851d624287f0120583a3d0e
# (245 runs: model size N, training FLOP C, final loss L; loss read from the
# figure's color scale, accurate to about 0.01). The five runs with the fewest
# training tokens per parameter are excluded, as in scaling-fit.py.
#
# Frontier points: one per IsoFLOP budget (6e18 to 3e21 FLOP). A budget's slice
# is the runs whose C is within a factor 1.2 of the budget; the reconstructed
# file also holds runs at other compute (a stray 1.3e22 FLOP run sits nearest
# the 3e21 budget), so the slice is taken on C itself, not on the nearest
# budget. The frontier loss of a slice is the minimum over the slice's range of
# model sizes of a parabola in log10 N fitted to its losses (Approach 2 of
# Hoffmann et al.): the lowest loss the budget reaches with the best model size.
#
# Fits: least squares on log L of each form to the frontier points at or below
# a cutoff budget; the points above the cutoff are held out and scored.
# Uncertainty: a bootstrap over runs. Each replicate resamples the runs within
# every slice with replacement, recomputes the nine frontier points, and refits
# both forms, so the band reflects the scatter of the runs behind each point.
# The file stores the 10th and 90th percentiles of each form's L(C) on a grid of
# budgets, and of each frontier point.
#
#   pip install numpy scipy        (run with numpy 2.5.3, scipy 1.18.1)
#   python tools/figure-data/compute-law-fit.py [path/to/svg_extracted_data.csv]

import csv
import io
import math
import sys
import urllib.request
from pathlib import Path

import numpy as np
from scipy.optimize import least_squares

COMMIT = "92258837425e1b5f2851d624287f0120583a3d0e"
URL = f"https://raw.githubusercontent.com/epoch-research/analyzing-chinchilla/{COMMIT}/data/svg_extracted_data.csv"
OUT = Path(__file__).resolve().parents[2] / "app" / "src" / "figures" / "data" / "compute-law-frontier.ts"

BUDGETS = [6e18, 1e19, 3e19, 6e19, 1e20, 3e20, 6e20, 1e21, 3e21]
TOL = 1.2  # a run belongs to a budget's slice when C is within this factor of it
CUTOFFS = [4, 5, 6, 7, 8]  # indices into BUDGETS: fit on budgets[0..k]
BOOTSTRAPS = 1000
SEED = 20200123
GRID = [18 + 0.125 * i for i in range(57)]  # log10 C from 18 to 25
C0 = 1e18  # compute is scaled by C0 inside the fits for conditioning


def load(path):
    text = Path(path).read_text() if path else urllib.request.urlopen(URL).read().decode()
    rows = list(csv.DictReader(io.StringIO(text)))
    N = np.array([float(r["Model Size"]) for r in rows])
    C = np.array([float(r["Training FLOP"]) for r in rows])
    L = np.array([float(r["loss"]) for r in rows])
    return N, C, L


# Lowest loss of a parabola in log10 N over the slice's range of model sizes.
def frontier_loss(lN, L):
    a, b, c = np.polyfit(lN, L, 2)
    lo, hi = lN.min(), lN.max()
    xs = [lo, hi]
    if a > 0:
        xs.append(min(max(-b / (2 * a), lo), hi))
    return min(a * x * x + b * x + c for x in xs), (-b / (2 * a) if a > 0 else float("nan"))


def fit_pure(c, L):
    # log L = log A - alpha log c: ordinary least squares.
    slope, icpt = np.polyfit(np.log(c), np.log(L), 1)
    return 0.0, math.exp(icpt), -slope


def fit_floor(c, L):
    lc, lL = np.log(c), np.log(L)
    top = L.min() - 1e-3

    def resid(q):
        linf, la, alpha = q
        return np.log(linf + np.exp(la - alpha * lc)) - lL

    best = None
    # Starting points from the profile over L_inf: a line through log(L - L_inf).
    for linf in np.linspace(0, top, 12):
        slope, icpt = np.polyfit(lc, np.log(L - linf), 1)
        x0 = [linf, icpt, max(1e-4, -slope)]
        r = least_squares(resid, x0, bounds=([0, -50, 1e-4], [top, 50, 3]))
        if best is None or r.cost < best.cost:
            best = r
    linf, la, alpha = best.x
    return float(linf), math.exp(la), float(alpha)


def predict(f, c):
    linf, A, alpha = f
    return linf + A * c ** (-alpha)


# A in the units of C itself (FLOP): A_c c^-alpha = A_c (C / C0)^-alpha.
def in_flop(f):
    linf, A, alpha = f
    return linf, A * C0 ** alpha, alpha


def g(v, digits=4):
    return float(f"{v:.{digits}g}")


def main():
    N, C, L = load(sys.argv[1] if len(sys.argv) > 1 else None)
    D = C / (6 * N)
    keep = np.ones(len(N), bool)
    keep[np.argsort(D / N)[:5]] = False
    slices = [np.where(keep & (np.abs(np.log10(C / b)) < math.log10(TOL)))[0] for b in BUDGETS]
    lN = np.log10(N)
    budgets = np.array(BUDGETS) / C0
    grid = 10 ** np.array(GRID) / C0

    points = [frontier_loss(lN[s], L[s]) for s in slices]
    front = np.array([p[0] for p in points])

    rng = np.random.default_rng(SEED)
    boot_front = []
    for _ in range(BOOTSTRAPS):
        row = []
        for s in slices:
            r = rng.choice(s, len(s), replace=True)
            row.append(frontier_loss(lN[r], L[r])[0])
        boot_front.append(row)
    boot_front = np.array(boot_front)

    fits = []
    for k in CUTOFFS:
        inside = slice(0, k + 1)
        entry = {"budget": BUDGETS[k], "points": k + 1}
        for form, fitter in (("pure", fit_pure), ("floor", fit_floor)):
            f = fitter(budgets[inside], front[inside])
            err_in = float(np.median(np.abs(predict(f, budgets[inside]) - front[inside])))
            held = front[k + 1:]
            err_out = float(np.median(np.abs(predict(f, budgets[k + 1:]) - held))) if len(held) else None
            curves = np.array([predict(fitter(budgets[inside], b[inside]), grid) for b in boot_front])
            linf, A, alpha = in_flop(f)
            entry[form] = {
                "Linf": g(linf, 4), "A": g(A, 5), "alpha": g(alpha, 4),
                "errFit": g(err_in, 2), "errHeld": g(err_out, 2) if err_out is not None else None,
                "lo": [g(v) for v in np.percentile(curves, 10, axis=0)],
                "hi": [g(v) for v in np.percentile(curves, 90, axis=0)],
            }
            print(f"fit <= {BUDGETS[k]:.0e} {form:5s}: L_inf={linf:.3f} A={A:.4g} alpha={alpha:.4f}, "
                  f"median |error| {err_in:.4f} fitted, {err_out if err_out is None else round(err_out, 4)} held out, "
                  f"L(5.76e23) = {predict(f, 5.76e23 / C0):.3f}", file=sys.stderr)
        fits.append(entry)

    rows = []
    for i, (s, (lf, xv)) in enumerate(zip(slices, points)):
        rows.append({
            "C": BUDGETS[i], "L": g(lf, 4),
            "lo": g(float(np.percentile(boot_front[:, i], 10)), 4),
            "hi": g(float(np.percentile(boot_front[:, i], 90)), 4),
            "runs": int(len(s)), "N": g(10 ** xv, 3),
        })
        print(f"budget {BUDGETS[i]:.0e}: {len(s)} runs, frontier L = {lf:.4f} at N = {10 ** xv:.3g}", file=sys.stderr)

    def js(v):
        if v is None:
            return "null"
        if isinstance(v, int):
            return str(v)
        m, _, e = f"{v:.6g}".partition("e")
        return m + ("e" + str(int(e)) if e else "")

    def obj(d):
        return "{ " + ", ".join(
            f"{k}: [{', '.join(js(x) for x in v)}]" if isinstance(v, list)
            else f"{k}: {obj(v)}" if isinstance(v, dict)
            else f"{k}: {js(v)}" for k, v in d.items()) + " }"

    lines = [
        "// Generated by tools/figure-data/compute-law-fit.py; do not edit by hand.",
        "//",
        "// POINTS: the compute frontier of the Chinchilla runs reconstructed by",
        "// Besiroglu et al. (2024) from Figure 4 of Hoffmann et al. (2022),",
        f"// epoch-research/analyzing-chinchilla data/svg_extracted_data.csv at commit {COMMIT[:7]}.",
        f"// One point per IsoFLOP budget C: the runs with training compute within a",
        f"// factor {TOL} of C, and the lowest loss L of a parabola in log10 N fitted",
        "// to them (Approach 2 of Hoffmann et al.), at the model size N. lo and hi are",
        f"// the 10th and 90th percentiles over {BOOTSTRAPS} bootstrap resamples of the runs.",
        "//",
        "// FITS: L(C) = Linf + A C^-alpha with C in FLOP, least squares on log L to",
        "// the points at or below `budget`: `pure` fixes Linf = 0 (the form of Kaplan",
        "// et al. 2020), `floor` fits Linf. errFit and errHeld are median absolute",
        "// errors on the fitted and held-out points; lo and hi are the 10th and 90th",
        "// percentiles of the fitted L(C) at log10 C = GRID over the same bootstrap.",
        "",
        f"export const GRID = [{', '.join(js(v) for v in GRID)}] as const;",
        "",
        "export interface FrontierPoint { C: number; L: number; lo: number; hi: number; runs: number; N: number }",
        "export const POINTS: readonly FrontierPoint[] = [",
        *[f"  {obj(r)}," for r in rows],
        "];",
        "",
        "export interface FormFit { Linf: number; A: number; alpha: number; errFit: number; errHeld: number | null; lo: number[]; hi: number[] }",
        "export interface CutoffFit { budget: number; points: number; pure: FormFit; floor: FormFit }",
        "export const FITS: readonly CutoffFit[] = [",
        *[f"  {obj(f)}," for f in fits],
        "];",
    ]
    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text("\n".join(lines) + "\n")
    print(f"wrote {OUT}", file=sys.stderr)


if __name__ == "__main__":
    main()
